src/vr/journal.zig

```diff
@@ -489,129 +729,241 @@ pub const Journal = struct {
         };
     }
 
-    /// Writes headers between `op_min` and `op_max` (both inclusive).
-    fn write_headers_between(self: *Journal, op_min: u64, op_max: u64) void {
-        // TODO Snapshots
-        assert(op_min <= op_max);
+    /// Attempt to lock the in-memory sector containing the header being written.
+    /// If the sector is already locked, add this write to the wait queue.
+    fn write_prepare_header(self: *Journal, write: *Journal.Write) void {
+        assert(!write.header_sector_locked);
+        assert(write.header_sector_next == null);
+
+        var it = self.writes.iterate();
+        while (it.next()) |other| {
+            if (other == write) continue;
+            if (!other.header_sector_locked) continue;
+
+            if (other.header_sector_same(write)) {
+                write.header_sector_next = other.header_sector_next;
+                other.header_sector_next = write;
+                return;
+            }
+        }
 
-        const offset = Journal.sector_floor(op_min * @sizeOf(Header));
-        const len = Journal.sector_ceil((op_max - op_min + 1) * @sizeOf(Header));
-        assert(len > 0);
+        write.header_sector_locked = true;
+        self.write_prepare_on_lock_header_sector(write);
+    }
 
-        // We must acquire the concurrent range using the sector offset and len:
-        // Different headers may share the same sector without any op_min or op_max overlap.
-        // TODO Use a callback to acquire the range instead of suspend/resume:
-        var range = Range{ .offset = offset, .len = len };
-        self.writing_headers.acquire(&range);
-        defer self.writing_headers.release(&range);
+    /// Release the lock held by a write on an in-memory header sector and pass
+    /// it to a waiting Write, if any.
+    fn write_prepare_unlock_header_sector(self: *Journal, write: *Journal.Write) void {
+        assert(write.header_sector_locked);
+        write.header_sector_locked = false;
+
+        // Unlike the ranges of physical memory we lock when writing to disk,
+        // these header sector locks are always an exact match, so there's no
+        // need to re-check the waiting writes against all other writes.
+        if (write.header_sector_next) |waiting| {
+            write.header_sector_next = null;
+
+            assert(waiting.header_sector_locked == false);
+            waiting.header_sector_locked = true;
+            self.write_prepare_on_lock_header_sector(waiting);
+        }
+        assert(write.header_sector_next == null);
+    }
+
+    fn write_prepare_on_lock_header_sector(self: *Journal, write: *Write) void {
+        assert(write.header_sector_locked);
```

Comment:
```suggestion
        assert(write.header_sector_locked);
        assert(write.header_sector_next == null);
```

Reply:
Actually, it might not be null here. We use the linked list slightly differently and more efficiently for headers since they are always an exact match, unlike the sector ranges below. In particular, we don't need to requeue every write that was waiting on a given write to complete. Instead, we just start the first waiting write in the wait list and leave the rest of the list intact.

Reply:
Great, I had actually made some similar commit suggestions to assert null for next pointers for sector ranges, before realizing these would break, and before deleting those commit suggestions.